packages/core/src/Synchronizer.ts

```diff
@@ -176,80 +215,68 @@ export class Synchronizer extends EventEmitter {
-            throw new Error(
-                '@unirep/core:Synchronizer: failed to fetch genesis event'
-            )
+        let events: ethers.Event[] = []
+        if (this._syncAll) {
+            const filter = this.unirepContract.filters.AttesterSignedUp()
+            events = await this.unirepContract.queryFilter(filter)
+        } else if (this._attesterId.length) {
+            for (let id of this._attesterId) {
+                const filter = this.unirepContract.filters.AttesterSignedUp(id)
+                const event = await this.unirepContract.queryFilter(filter)
+                // const events = await this.unirepContract.queryFilter(filter)
+                if (event.length === 0) {
+                    throw new Error(
+                        `@unirep/core:Synchronizer: failed to fetch genesis event of attester ${id.toString()}`
+                    )
+                }
+                if (event.length > 1) {
+                    throw new Error(
+                        '@unirep/core:Synchronizer: multiple genesis events'
+                    )
+                }
+                events.push(event[0])
+            }
         }
-        if (events.length > 1) {
-            throw new Error(
-                '@unirep/core:Synchronizer: multiple genesis events'
+
+        for (let event of events) {
+            const decodedData = this.unirepContract.interface.decodeEventLog(
+                'AttesterSignedUp',
+                event.data,
+                event.topics
             )
-        }
-        const [event] = events
-        const decodedData = this.unirepContract.interface.decodeEventLog(
-            'AttesterSignedUp',
-            event.data,
-            event.topics
-        )
-        const { timestamp, epochLength } = decodedData
-        this.settings.startTimestamp = Number(timestamp)
-        this.settings.epochLength = Number(epochLength)
-        const syncStartBlock = event.blockNumber - 1
+            const { timestamp, epochLength, attesterId } = decodedData
+            this._attesterSettings[toDecString(attesterId)] = {
+                startTimestamp: Number(timestamp),
+                epochLength: Number(epochLength),
+            }
+            if (
+                this._syncAll &&
+                !this.attesterExist(attesterId) &&
+                BigInt(attesterId) !== BigInt(0)
+            ) {
+                this._attesterId.push(BigInt(attesterId))
+            }
+            const syncStartBlock = event.blockNumber - 1
 
-        await this._db.upsert('SynchronizerState', {
-            where: {
-                attesterId: this.attesterId.toString(),
-            },
-            create: {
-                attesterId: this.attesterId.toString(),
-                latestCompleteBlock: syncStartBlock,
-            },
-            update: {},
-        })
+            await this._db.upsert('SynchronizerState', {
+                where: {
+                    latestCompleteBlock: {
```

Comment:
We need to consider the case where a database is used by multiple synchronizers.

e.g. We start a synchronizer syncing attesters `0x1`, `0x2`, `0x3`. Then we stop it and start it again but now syncing `0xa`, `0xb`, `0xc` (using the same database). Does this work with the current scheme?

The single attester version was designed to allow 1 database to be used by many synchronizers.

Reply:
Yeah right
it should be with attester IDs
[8645c48](https://github.com/Unirep/Unirep/pull/367/commits/8645c48a6d9482179dba6df2844b02e358ee9f70)